sycl/plugins/unified_runtime/ur/adapters/cuda/usm.hpp

```diff
@@ -0,0 +1,130 @@
+//===--------- usm.hpp - CUDA Adapter --------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===-----------------------------------------------------------------===//
+
+#include "common.hpp"
+
+#include <umf_helpers.hpp>
+#include <umf_pools/disjoint_pool_config_parser.hpp>
+
+usm::DisjointPoolAllConfigs InitializeDisjointPoolConfig();
+
+struct ur_usm_pool_handle_t_ {
+  std::atomic_uint32_t RefCount = 1;
+
+  ur_context_handle_t Context = nullptr;
+
+  usm::DisjointPoolAllConfigs DisjointPoolConfigs =
+      usm::DisjointPoolAllConfigs();
+
+  umf::pool_unique_handle_t DeviceMemPool;
```

Comment:
I know that for L0 we have pools for Device, Host, Shared and SharedReadOnly. Do you think it would make sense to do the same for CUDA?

Also, as far as I understand CUDA is already doing some memory pooling internally, right? I also see there is `memoryPool` abstraction in CUDA as well, would it make sense to use it in this implementation? https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MALLOC__ASYNC.html#group__CUDA__MALLOC__ASYNC_1g8aa4c143dbc20293659cd883232b95f2

Reply:
The only one I left out was SharedReadOnly, cuda doesn't have a way to support that really so all shared allocations can use the same pool.

I totally didn't notice you can pool device allocations explicitly like that with cuda, I'll have a go at integrating it into the device pool now

Reply:
Oh, I see, then I think you're right, it's probably not worth using it. Btw, do you know if CUDA does any pooling automatically?

Reply:
Nvidia don't seem eager to document to that level of detail. I had a look and found this https://saturncloud.io/blog/is-it-possible-to-solve-cuda-memory-fragmentation-issue/ which claims cuda managed memory (which is how we implement the UR `Shared` allocations) has some kind of built in pooling